Add portable support

n0routine opened the pull request:
/claim #888

#### Brief overview of PR changes/additions

This PR:
- refactors Mudlet initialisation code to use a single source of truth for config directory paths
- change QSettings storage to an `IniFormat` file in the config directory and adds migration code from old formats
- adds support for a `<executable_dir>/portable.txt` marker which sets the config directory to `<executable_dir>/portable`

#### Motivation for adding to Mudlet

This PR teaches Mudlet to store data portably.

#### Usage Instructions

If you're interested in trying out this PR, you can download the CI builds from the **add-deployment-links** bot below.

There are currently two ways to enable portable mode for Mudlet (in order of importance):
- You can create an empty `portable.txt` file in the same folder as the Mudlet executable (or appimage etc.)
  - This will tell Mudlet to use a folder named `portable` (in the same folder as the executable) for its data
- You can create an `~/.config/mudlet/portable.txt` file with its contents being a path on your filesystem
  - This will tell Mudlet to use the path written in the file as the folder for its data
    -  e.g. the contents could be `/mount/media/flashdrive/mudlet_data` or `D:\games\portable\mudlet_data`
  - The path can be relative, in which case it will be interpreted relative to the Mudlet executable's folder
  - Mudlet will create _exactly_ one folder, that is, the last part of the given path, if it doesn't exist already. At least everything up to its parent folder must exist already or it will lead to an error.
    - e.g. if given `D:\games\portable\mudlet_data`, at least `D:\games\portable` must be an existing folder
    - This is to avoid taking unintended input. If you see this error but this actually is what you want, just create those folders manually

Any errors will result in the issue being printed to stderr and the program terminating. You probably won't see the error outputs if launched from GUI so it's recommended to start Mudlet from the terminal.

Ofc when you first launch Mudlet in portable mode, it will start with a new clean config in the respective folder, just like a new install. If you wish to migrate your existing config data to be portable:
- You need to launch this build at least once (without any `portable.txt`)
  - This is because it needs to migrate config files from the old format to the new, portable-friendly one
- Then you can just copy/move your default data directory `~/.config/mudlet` to wherever you want and use one of the above `portable.txt`s to point Mudlet to that path.

n0routine committed:
mudlet::getMudletPath: factor out confPath

This makes no functional changes.

n0routine committed:
mudlet::confPath: add member to hold base config path

Add new method setupConfig() to initialise member.
Also change getMudletPath() to use confPath.

Move mudlet initialisation above, before any files have to be read.

This makes no fucntional changes.

n0routine committed:
mudlet::init: move all initialisation from constructor

This is necessary because the config path needs to be initialised
much earlier in main(), before any file operations take place.

This makes no functional changes.

n0routine committed:
mudlet::setupConfig: use Mudlet.ini to store changes

Also add code to migrate from the old, default paths.

n0routine committed:
mudlet::getQSettings: turn into actual getter

mpSettings is only created once in setupConfig().

Change all sites that access the config to call this getter.

This makes no functional changes.

n0routine committed:
mudlet::setupConfig: add portable support

If a `portable.txt` is found in the same directory as the executable,
all the data will be stored in a `<executable_directory>/portable`.

n0routine commented:
Yet to add 2nd fallback path and custom data paths. The fun parts are over, just boring changes to make now, going to do that when I get back.

I've tested and the behaviour seems fine on Linux, including migration from the old config. I can test on windows in some time. Unfortunately, I don't have a macOS machine to test it on.

vadi2 commented:
Very quick crab! :)

I can test on macOS for you. Mind filling out the PR template? It helps us stay organised.

n0routine commented:
@vadi2 updated the PR body.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -691,16 +693,43 @@ mudlet::mudlet()
+}
+
+void mudlet::setupConfig() {
+    if (QFile::exists(qsl("%1/portable.txt").arg(QCoreApplication::applicationDirPath()))) {
```
I believe this should work fine on Windows. Not sure about macOS right now.

It doesn't work in Linux when running via an AppImage though (our official distribution method) since those get extracted to a temporary directory prior to running:

> access("/tmp/.mount_MudletSsu3Rv/portable.txt", F_OK) = -1 ENOENT (No such file or directory)

To solve this, have a look at the code [here](https://github.com/Mudlet/Mudlet/blob/development/src/updater.cpp#L291-L294) which figures out the location of the AppImage.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -130,6 +130,10 @@ bool TConsoleMonitor::eventFilter(QObject* obj, QEvent* event)
 
 mudlet::mudlet()
 : QMainWindow()
+{
```
Brilliant solution! I love how clean the diff is.

Could you add a comment here explaining the "why" of having an empty constructor is a good idea?

vadi2 reviewed (commented):
Some early feedback - will review more later on :)

n0routine committed:
mudlet::setupConfig: look for ~/.config/mudlet/portable.txt marker

This file will be read to get a path which will be set to the
config directory.

n0routine commented:
@vadi2 addressed the reviews.

Added the 2nd path and reading from the file... but unsure of some semantics:
- For the executable dir marker, an empty portable.txt should probably mean the nice default of `./portable`, and if it has a path, should set it to that path. For the `~/.config` marker with a path inside obviously makes sense. But what about an empty markerfile? Do we then set the config to something like `~/.config/mudlet/portable`? Should this case be an error?
- Do we need to validate the path in some way? Does the directory need to exist? Does everything except the last component need to exist? It obviously shouldn't be an existing file, I guess.
- What do we do with the above errors?

vadi2 commented:
From the point of view of explaining the feature to others, and helping them debug it, I would keep the behaviors of both files the same to keep things as simple as possible. This would mean that you'd have to fill a path in `portable.txt` - it might take 30s more to do, but could save an hour of debugging.

I'll think on the other two points and get back to you.

n0routine commented:
> From the point of view of explaining the feature to others, and helping them debug it, I would keep the behaviors of both files the same to keep things as simple as possible. This would mean that you'd have to fill a path in `portable.txt` - it might take 30s more to do, but could save an hour of debugging.

I agree that we should try to keep the rules simple. That's why I was unsure, and wanted to discuss this.

Because without a simple case expressing `./portable`, I don't think it counts as _truly_ portable, not in the way users would expect normally. You can do stuff like pointing different installations to the same data. But you can't simply plop an executable + data dir onto a flash drive and launch it anywhere: it'll have a different mount path everywhere.

One workaround we could use is accepting relative paths i.e. users can write `./portable` in the marker files. Note that this is not something that works correctly in the current implementation: any relative paths are naively resolved against the working directory, which could be anything. We would need to add code to explicitly resolve relative paths against the executable/markerfile (whichever makes sense, IG). Not sure if there are more wrinkes to this approach.

Also IMO the empty file option is a nice way for an easy out for 90% of users that just want the simple portable behavior. The custom path option is going to be used by fewer users and it has more chance for user error (e.g. putting in an incorrect path, or something they didn't want to).

vadi2 commented:
Quite a lot of Mudlet folks want Mudlet to be portable only to be able to sync their saves across cloud systems, e.g. https://forums.mudlet.org/viewtopic.php?f=8&t=1731 was around from Mudlets beginning. There's no in having it installed on the system otherwise. I didn't mention that we'll enough in the original issue.

That said the best way to know what is right is to actually try it. Let's try the empty portable.txt to mean the data is in `./mudlet` (that way you can copy .config/mudlet without having an additional rename step), and allow for relative paths as well. Data doesn't have to be present already, and if there are any issues (e.g. opening files or anything else), print them with a qWarning(). 

Then let's get it out to folks for testing!

n0routine commented:
> That said the best way to know what is right is to actually try it.

That's true!

> Let's try the empty portable.txt to mean the data is in `./mudlet`

I... think that could be problematic if the executable is `mudlet` as well. How about `./mudlet_data` or something? It would still need a rename but would be more obvious/descriptive, I guess.

> and allow for relative paths as well

Gotcha.

> Data doesn't have to be present already

Well, yeah, but I meant, like, if the user specifies `/home/user/data/games/mudlet`:
 - Do we require that a directory `/home/user/data/games/mudlet` already exist? (this is probably too strict)
 - Do we require that a directory `/home/user/data/games` already exist? i.e. we require everything up to the last component exist, effectively preventing a `mkdir -p` (not necessary but IMO this pretty reasonable restriction would prevent a lot of mistaken user inputs, by making sure the user knows what they're doing, instead of accepting basically any string)

The one obvious restriction that occurs to me is that it shouldn't be an existing file.

> and if there are any issues (e.g. opening files or anything else), print them with a qWarning().

And then should we terminate or continue with a default path? IMO if the user asked for something and we can't do it then we should loudly terminate so they know about it and then can fix it.

Can users who launched the app from a GUI see `qWarning`s?

wiploo commented:
Hi, I'm interested in this PR, but I can't understand how can i use it? Can you explain in a quick-start-tutorial?

vadi2 commented:
> think that could be problematic if the executable is mudlet as well. 

One is a directory and the other is a file, what kind of issues are you envisioning?

> Do we require that a directory /home/user/data/games/mudlet already exist?  

Lets have the code create it as needed.

> Do we require that a directory /home/user/data/games already exist

I personally don't see an issue with that as long as the location can be written to. If the location is wrong, the user will work that out for themselves and fix it - but I don't mind it too much either way, as it's hard to tell w/o testing.

> The one obvious restriction that occurs to me is that it shouldn't be an existing file.

Yep, but a directory is OK.

> And then should we terminate or continue with a default path? IMO if the user asked for something and we can't do it then we should loudly terminate so they know about it and then can fix it.

Lets think about it on this one. In my experience of supporting all kinds of users for two decades, throwing errors at users works great for the technically inclined ones, but not so much for the non-technically inclined. I'll check in with the rest of the team on this.

n0routine commented:
> One is a directory and the other is a file, what kind of issues are you envisioning

Well, we can't have the executable overwriting itself now, can we?

n0routine commented:
> Hi, I'm interested in this PR, but I can't understand how can i use it? Can you explain in a quick-start-tutorial?

@wiploo these changes basically allow you to make Mudlet use a custom path for storing its config data.

There are currently two portable "modes":
- You can create an empty `portable.txt` file in the same directory as the Mudlet executable (/appimage/*)
  - This makes Mudlet use a `portable` folder in the same directory as the executable for its data
- You can create an `~/.config/mudlet/portable.txt` file with its contents being a path on your filesystem
  - This makes Mudlet use the path written in the file for its data directory
  -  e.g. the contents could be `/mount/media/flashdrive/mudlet_data` or `D:\games\mudlet_data`

Ofc when you first launch Mudlet in portable mode, it will start with a new clean config in the respective folder, just like a new install.

Keep in mind that the implementation is vary naive and error handling of unexpected/unwanted situations is non-existent as of yet.

As for migrating your existing config data to be portable:
- You need to launch this build at least once (without a `portable.txt`) because it needs to migrate the old config format to the new format which is portable-friendly.
- Then you can just move your default data directory `~/.config/mudlet` to whatever location you want and use one of the above `portable.txt`s to point Mudlet to that path.

n0routine commented:
Do note that some of this behaviour is subject to change as discussed in the conversation with @vadi2 above. As @vadi2 also said, we'd be glad to take input on what users would consider useful/sensible/non-confusing behaviour about the lines of discussion above.

vadi2 commented:
The binary is called mudlet.exe on windows so the name won't clash, and Linux & macOS filesystems are clever enough to tell a file and a directory apart. Can you be more specific in your line of thinking?

n0routine commented:
> Linux & macOS filesystems are clever enough to tell a file and a directory apart

@vadi2 I'm... pretty sure that's not the case :p

vadi2 commented:
Yep, seems not. Alright, let's go with the well-known Dolphin behaviour here.

SlySven commented:
> > Linux & macOS filesystems are clever enough to tell a file and a directory apart
> 
> @vadi2 I'm... pretty sure that's not the case :p

Yeah - everything is a file - even a directory - it is just the contents that make it special...

n0routine commented:
So I implemented some of the changes we talked about earlier:

- Added basic validation for paths in the `~` markerfile
  - The file cannot be empty
    - Rationale: probably makes no sense
  - The path cannot be an existing file
    - Rationale: don't want to overwrite files
  - The path's parent must be an existing dir
    - Rationale:
      - We don't have to worry about handling and accurately reporting many errors like: path given `/I/J/K/L` but `/I/J` is an existing file. As long as `K` is a an existing directory, we're basically fine.
      - Makes it harder for the user to make mistakes: without this check the user could basically keyboard smash any string and Mudlet will happily `mkdir -p` it. By requiring the user to ensure themselves that `K` exists, we ensure that they know what they're doing and massively reduce the space of mistaken inputs we accept
- Added `qWarning`s for the validation errors, and a `qFatal` if any of them are tripped
  - Rationale: nothing in particular, the `qFatal` is just for testing purposes, can change it to whatever later
  - But IMO a validation error should be loud (whether it be a termination or error dialog) instead of silently continuing on
- Relative paths in the `~` markerfile are resolved against the executable directory
- Fixed some bugs about paths caused by my misunderstanding of QT file APIs
  - In my defense, some of them use very ambiguous language in the docs

vadi2 commented:
Looks awesome. Could you add the latest instructions to test to the top comment? We'll get folks to test out the binary builds available in this PR then.

n0routine commented:
@vadi2 added instructions to the pr body.

vadi2 commented:
Awesome. Added it to our Discord over at https://discord.com/channels/283581582550237184/427919962561052673/1268488023910252636 ([invite](https://discord.gg/kuYvMQ9))

n0routine commented:
Hi @vadi2,
any updates on this from testing?

I had intended to join your discord but then I realised my throwaway account on it is still blocked since they enforced number verification :P

vadi2 commented:
We had one tester who followed these steps to enable a portable setup:

1. Create portable.txt where mudlet.exe is
1. Copy mudlet folder and rename it to portable
1. Launch Mudlet

They then created a new profile, quit portable Mudlet, re-opened normal Mudlet and the new profile was visible there. I haven't double checked this yet, but it sounds like the new profile should _not_ appear in the normal Mudlet build.

vadi2 commented:
I haven't tried this yet, will soon, been dealing with other issues and letting others test this one out. I'll try it myself soonish.

By the way we released $1,200 worth of bounties more today: https://console.algora.io/org/Mudlet

n0routine commented:
> We had one tester who followed these steps to enable a portable setup:
> 
> 1. Create portable.txt where mudlet.exe is
> 2. Copy mudlet folder and rename it to portable
> 3. Launch Mudlet
> 
> They then created a new profile, quit portable Mudlet, re-opened normal Mudlet and the new profile was visible there.

Huh. That's weird.

`mudlet.exe` so I assume windows, but I tested the windows builds above and they seemed to respect both the markerfiles fine.

Unfortunately, I actually don't play MUDs :( so I didn't actually fiddle with a lot of non-trivial knobs, only stuff like languages, dark mode, etc. And they seemed to be isolated well enough.

Where (codewise) do we store profile data? If it's somewhere outside the flow of of `getMudletData()` and the handful of `QSettings` constructors I found with `git grep`, that site might've been missed by the current changes.

> By the way we released $1,200 worth of bounties more today: https://console.algora.io/org/Mudlet

Oh, neat, I'll be looking at them when I'm free.

vadi2 commented:
The Host class is the internal name for a profile - it stores triggers, aliases and so on and it's serialization is started by https://github.com/Mudlet/Mudlet/blob/d25db59bb1d1dc7916f3e25b7a3090f4275a58d7/src/Host.cpp#L901

n0routine commented:
Github code search shows a few uses of `Host::saveProfile`:

https://github.com/search?q=repo%3AMudlet%2FMudlet+saveProfile%28&type=code

Most of them seem to omit the first argument, making it get the default path from `getMudletPath`.

The one exception seems to be its lualand wrapper:

https://github.com/Mudlet/Mudlet/blob/d25db59bb1d1dc7916f3e25b7a3090f4275a58d7/src/TLuaInterpreter.cpp#L1318-L1327

Is it possible they're seeing this behaviour because of something they have set or installed in their lua context?

(aside: the lua interpreter run remote code sent by servers, no? I hope this part of the API isn't accessible to the remote code, it controlling the full path could get pretty nasty.)

vadi2 commented:
Not in the Lua context. Thinking about it more, profile creation is handled by  https://github.com/Mudlet/Mudlet/blob/d25db59bb1d1dc7916f3e25b7a3090f4275a58d7/src/dlgConnectionProfiles.cpp#L605 and the different elements are saved to disk using https://github.com/Mudlet/Mudlet/blob/d25db59bb1d1dc7916f3e25b7a3090f4275a58d7/src/dlgConnectionProfiles.cpp#L550-L574.

If you create a new profile in a portable Mudlet, does it end up where it should be in, the portable dir?

SlySven commented:
Re: https://github.com/Mudlet/Mudlet/pull/7375#issuecomment-2260117895
> There are currently two portable "modes":
> *    You can create an empty portable.txt file in the same directory as the Mudlet executable (/appimage/*)
        This makes Mudlet use a portable folder in the same directory as the executable for its data
> * ...

Isn't that going to awkward if that is done automagically by any code inserted into Mudlet for the GNU/Linux case - as **I *think*** running the AppImage executable unarchives itself to a temporary directory and proceeds to run the Mudlet executable contained therein - which will go away when the program ends - taking out any data the user has saved...

n0routine commented:
> Not in the Lua context.

Do you mean that this issue can't be because of the above function being called, or did I misunderstand something in the above mechanism?

> Thinking about it more, profile creation is handled by
> https://github.com/Mudlet/Mudlet/blob/d25db59bb1d1dc7916f3e25b7a3090f4275a58d7/src/dlgConnectionProfiles.cpp#L605
> 
> and the different elements are saved to disk using
> https://github.com/Mudlet/Mudlet/blob/d25db59bb1d1dc7916f3e25b7a3090f4275a58d7/src/dlgConnectionProfiles.cpp#L550-L574

Hm, at first look everything seems to be in order here and go through `getMudletPath()` like a good citizen.

> If you create a new profile in a portable Mudlet, does it end up where it should be in, the portable dir?

Yep, cannot seem to reproduce on my side. Tested on a windows machine. Profiles created in both the default data dir and the portable data dir seem to be isolated, confirmed by both in-game GUI and the `profiles/<profile>` directories being made.

Assuming a valid way of creating such profiles is selecting one of the servers on the "connect" screen, pressing copy and giving them a funny name.

n0routine commented:
Hm, if they a windows user, what they see as `portable.txt` might not actually be one, since windows hides extensions by default.

How are they testing whether their mudlet instance is in normal/portable mode? Does it create a portable directory (if they delete the one they copied manually)?

I had added a log message indicating the data directory being used, but it seems `qDebug` only outputs in debug builds (kinda duh in retrospect).

n0routine commented:
> Re: [#7375 (comment)](https://github.com/Mudlet/Mudlet/pull/7375#issuecomment-2260117895)
> 
> > There are currently two portable "modes":
> > 
> > * You can create an empty portable.txt file in the same directory as the Mudlet executable (/appimage/*)
> >   This makes Mudlet use a portable folder in the same directory as the executable for its data
> > * ...
> 
> Isn't that going to awkward if that is done automagically by any code inserted into Mudlet for the GNU/Linux case - as **I _think_** running the AppImage executable unarchives itself to a temporary directory and proceeds to run the Mudlet executable contained therein - which will go away when the program ends - taking out any data the user has saved...

@SlySven yeah, @vadi2 pointed this out in an earlier review, so in case of it being an appimage, the code considers "executable dir" to mean "the appimage-containing dir":

https://github.com/Mudlet/Mudlet/pull/7375/files#diff-a4fc43e16d0ada6ab826b586dbc8c8600ac838f7558a4fd7c7434db135f384e1R697-R705

vadi2 commented:
https://github.com/Mudlet/Mudlet/pull/7375#issuecomment-2273328330 this worked fine for me as well with an empty portable.txt.

I then added a path to it (just `C:\Users\vadi\Downloads\portable` as contents) but that didn't take hold, only the `./portable` was used. Does that work for you?

vadi2 commented:
Ah I see that the differences in behaviour is intended - but folks will find it confusing. It would be good if the file worked the same regardless of it's location.

n0routine commented:
> [#7375 (comment)](https://github.com/Mudlet/Mudlet/pull/7375#issuecomment-2273328330) this worked fine for me as well with an empty portable.txt.
> 
> I then added a path to it (just `C:\Users\vadi\Downloads\portable` as contents) but that didn't take hold, only the `./portable` was used. Does that work for you?

Oh, uh, heh, it seems the markerfile in the executable dir doesn't read the path inside:

https://github.com/Mudlet/Mudlet/pull/7375/files#diff-a4fc43e16d0ada6ab826b586dbc8c8600ac838f7558a4fd7c7434db135f384e1R760-R761

> Ah I see that the differences in behaviour is intended - but folks will find it confusing. It would be good if the file worked the same regardless of it's location.

I wouldn't say "intended", I guess it may be more accurate to say that I 'just' forgot to implement that :D 

It sounds a little funny after all the discussion about its semantics that day, but yeah lol.

But I did test the custom paths with the `.config` markerfile, so we can just copy that logic, only with the default fallback added.

SlySven commented on src/mudlet.cpp:
```diff
@@ -691,16 +694,92 @@ mudlet::mudlet()
+}
+
+static bool validateConfDir(QString& path) {
+    if (path.size() == 0) {
```
:information_source: We try to use `QString::isEmpty()` instead of checking the size as it more succinctly expresses the intent of the check: 
```suggestion
    if (path.isEmpty()) {
```

n0routine commented:
Also it occurs to me that there's ways that people could be confused when a "portable" instance might look like it shares data with the "normal" instance.

Currently, on every fresh launch we look for the native `QSettings` object and migrate data from it. If someone has the old native settings, every launch of a fresh instance, including any portable ones, would try to migrate from it, and inherit everything. That sounds confusing since the users would expect it to be a fresh install (and it is in every respect other than the settings).

Maybe we should `.clear()` the native settings after migration so that this process only happens once?

SlySven commented on src/mudlet.cpp:
```diff
@@ -691,16 +694,92 @@ mudlet::mudlet()
+    }
+    QFileInfo pathInfo(path);
+    if (pathInfo.isFile()) {
+        qWarning("WARN: specified portable data path is an existing file: %s", qPrintable(path));
```
Using `QWarning`/`QDebug`/`QInfo` for these messages may not be the best option here as they may not show up for all users - especially if they are not running Mudlet from a terminal window. For these particular cases it might be appropriate to use the normally discouraged by us (Core Devs) `QMessageBox` - as it is, this sort of problem is sufficient to stop the end-user using Mudlet in the manner they wish so having to stop and read a "modal" message could, I think, be tolerated. @vadi2 your thoughts on this?

SlySven replied:
Also, these messages are not going to be translated so could be problematic for non-English users...

vadi2 replied:
Having had supported all sorts of users for two decades now, I reckon this warning is OK and better than a pop-up.

In case someone is active trying to make Mudlet portable work and it's not working, we can advise them to launch from the terminal to see any issues.

In case someone non-technical isn't trying to make Mudlet portable work though, and they get blocking pop ups, it's always a bit of a disaster as people get confused / frustrated / upset and don't know what to do.

So this approach has the least problems.

SlySven commented on src/mudlet.cpp:
```diff
@@ -691,16 +694,92 @@ mudlet::mudlet()
+}
+
+static void migrateConfig(QSettings& settings) {
+    if (settings.contains("pos")) {
```
It might have been picked up elsewhere but please don't use raw string literals in the C++ code:

* Either - if they are only used internally and not visible to the end-user or are output to the user from the Lua sub-system:  wrap them in `qsl(...)` (actually our macro for the more to type `QStringLiteral(...)`). This case applies here!
* Or - if they are presented to the end-user in the UI - wrap them in `tr(...)` (actually `QObject::tr(...)`) or for use in source files/classes not including the `QObject` or `Q_DECLARE_TR_FUNCTIONS` in their header files or - like `main.cpp` - not having a header file - in `QApplication::translate(...)`.

SlySven commented on src/mudlet.cpp:
```diff
@@ -691,16 +694,92 @@ mudlet::mudlet()
+    return true;
+}
+
+static void migrateConfig(QSettings& settings) {
```
Our coding style puts the opening `{` of a method/function at column 1 on the next line. Please adjust this (it is the one thing that `clang-format` doesn't pick up on IIRC)...

SlySven commented:
> Maybe we should .clear() the native settings after migration so that this process only happens once?

That might bite us - or rather users of - say the PTBs or testers - if they switch between Mudlet versions that know about "portable" support and those that do not!

vadi2 commented:
> Maybe we should .clear() the native settings after migration so that this process only happens once?

That makes sense, at the same time, we'd need a way to migrate the settings back somehow if one wants to undo portability

n0routine commented:
> That might bite us - or rather users of - say the PTBs or testers - if they switch between Mudlet versions that know about "portable" support and those that do not!

> That makes sense, at the same time, we'd need a way to migrate the settings back somehow if one wants to undo portability

True, it does bite, but I suppose the question is whether it bites more or less than the problem we'd have with the current code: people seeing supposedly fresh installs inherit everything from their old installs. (I would personally guess the answer is more but only slightly so).

Can we have any other workarounds? Maybe adding a `"has_been_migrated_from"` key and checking for that?

vadi2 commented:
I reckon we can have both, a clean migration to portable and a migration off portable. What do you think?

n0routine commented:
I wouldn't think of this problem as being related to "portable" mudlet.

The actual (irreversible) migration we're making here is the one from `NativeFormat` to `Mudlet.ini`, for _everyone_.

Once we have that uniform representation for the config, portable "migrations" are trivial: it's just copying the folders to-and-fro.

The problem comes from us adding the config migration change _along with_ adding a very specific feature that allows multiple independent instances to coexist. So suddenly every instance wants to migrate, while we only want one of them to.

i.e. this is the kind of change that would be handled in a corporate context by doing the migration as step A, and adding the actual feature as step B 5 years later :P

n0routine committed:
mudlet::setupConfig: read contents of execdir portable.txt

If empty, fall back to the default values of "./portable"

n0routine commented:
@SlySven addressed the reviews. Well, those that had consensus, anyway.

@vadi2 made the execdir markerfile read its contents (but unlike the home markerfile, being filled in is optional).

vadi2 commented:
Right. Developers frequently go back and forth between older and newer versions, and a one-off migration would be rather inconvenient. Lets keep it that way until we do hear from the players that it is a troublesome issue for them.

vadi2 commented:
Video of the failing test case - https://streamable.com/ohn78e

vadi2 commented:
They were following this concise method:

1. Create portable.txt where mudlet.exe is
2. Copy mudlet folder and rename it to portable 
3. Launch Mudlet

What does work for them is this very long method:

```1. unzipped contents into a folder on a blank usb stick into a folder called mudlet porta
2. ran the exe once
3. closed that mudlet
4. created a blank portable.txt in the dir with the mudlet exe
5. ran the mudlet exe again, portable dir was created.
6. closed mudlet
7. deleted portable folder
8. copied mudlet folder where all my profiles n stuff are to the usb stick in the same folder as mudlet exe
9. renamed mudlet folder to portable
10. opened the mudlet exe
11. everything is there as it should be!```

n0routine commented:
> They were following this concise method:
> 
>     1. Create portable.txt where mudlet.exe is
> 
>     2. Copy mudlet folder and rename it to portable
> 
>     3. Launch Mudlet
> 
> 
> What does work for them is this very long method:
> 
> ```
> 2. ran the exe once
> 3. closed that mudlet
> 4. created a blank portable.txt in the dir with the mudlet exe
> 5. ran the mudlet exe again, portable dir was created.
> 6. closed mudlet
> 7. deleted portable folder
> 8. copied mudlet folder where all my profiles n stuff are to the usb stick in the same folder as mudlet exe
> 9. renamed mudlet folder to portable
> 10. opened the mudlet exe
> 11. everything is there as it should be!```
> ```

So, uh, mentally diffing those methods, the difference is... that it works when they let Mudlet create the portable data folder and then replace it, but it doesn't when they copy the portable folder right away?

That's bizzare.

> Video of the failing test case - https://streamable.com/ohn78e

I've watched this a few times and can't see them doing anything wrong.

vadi2 commented:
In the video you see them launching the normal Mudlet build (with a yellow icon, no portable support) and somehow the new profile is showing up in it. 

I'll get the user to check the contents of ~/.config/mudlet/profiles and ./portable/profiles, see where the profile is actually created.

n0routine commented:
One thing I did notice was that they launched Mudlet through the start menu. Presumably that's the installed latest stable release, right? It's... sort of... impossible that it could be reading profiles from the portable data directory on the USB? The only logical conclusion would be that the Mudlet on their USB is creating profiles in the default data dir.

Are we sure portable mode is being triggered at all? It feels like we're trying to deduce an answer through shadows on a curtain.
I noticed that the CI builds are actually debug builds, it's just that for some strange reason `qDebug` messages don't show up if you launch the exe from powershell. I think they did show up when I launched it from git ([for windows](https://github.com/git-for-windows/git/releases)) bash. It could shed a lot of light if they could just answer the question of, in each of the above cases they describe, what data directory _Mudlet_ says it was actually using.

Also one thing they could try is:
- Make a `portable.txt`
- Don't copy any existing data into the `portable` folder (basically we want a fresh launch)
- Launch Mudlet

What do they see? Do they see a fresh launch or all of their existing settings? If they create a profile in that fresh data, does that show up in their main Mudlet as well?

n0routine commented:
> In the video you see them launching the normal Mudlet build (with a yellow icon, no portable support) and somehow the new profile is showing up in it.

Yeah I noticed the issue, I just meant that as far as I can see what they did should work, and they... shouldn't be seeing what they're seeing.

vadi2 reviewed (approved):
Sorted - windows was confusing them into creating portable.txt.txt.

n0routine commented:
Whoops. Did sorta call it. It was confusing though because in the video at least some of the files did have extensions, but windows do windoes.

vadi2 commented:
Looks like others are good with it too - merging :partying_face:
